File: c/tests/test_rlnc.c
// SPDX-License-Identifier: MIT
// Unit tests for rlnc.c — GF(2⁸) arithmetic, encoder, decoder, codec.

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "aether/rlnc.h"
#include "aether/transport.h"

// ── Helpers ───────────────────────────────────────────────────────────────────

/** Allocate a k×symbol_size source symbol array filled with predictable bytes. */
static uint8_t **make_source(int k, int sym_size) {
    uint8_t **src = (uint8_t **)malloc((size_t)k * sizeof(uint8_t *));
    assert(src != NULL);
    for (int i = 0; i < k; i++) {
        src[i] = (uint8_t *)malloc((size_t)sym_size);
        assert(src[i] != NULL);
        for (int j = 0; j < sym_size; j++) {
            src[i][j] = (uint8_t)((i * sym_size + j) & 0xFF);
        }
    }
    return src;
}

static void free_source(uint8_t **src, int k) {
    for (int i = 0; i < k; i++) free(src[i]);
    free(src);
}

// ── GF(2⁸) arithmetic ────────────────────────────────────────────────────────

static void test_gf256_add_is_xor(void) {
    aether_gf256_init();
    assert(aether_gf256_add(0xAB, 0xCD) == (uint8_t)(0xAB ^ 0xCD));
    assert(aether_gf256_add(0x00, 0xFF) == 0xFF);
    assert(aether_gf256_add(0xFF, 0xFF) == 0x00);
    printf("  PASS test_gf256_add_is_xor\n");
}

static void test_gf256_mul_by_zero(void) {
    aether_gf256_init();
    for (int a = 0; a < 256; a++) {
        assert(aether_gf256_mul((uint8_t)a, 0) == 0);
        assert(aether_gf256_mul(0, (uint8_t)a) == 0);
    }
    printf("  PASS test_gf256_mul_by_zero\n");
}

static void test_gf256_mul_by_one(void) {
    aether_gf256_init();
    for (int a = 0; a < 256; a++) {
        assert(aether_gf256_mul((uint8_t)a, 1) == (uint8_t)a);
        assert(aether_gf256_mul(1, (uint8_t)a) == (uint8_t)a);
    }
    printf("  PASS test_gf256_mul_by_one\n");
}

static void test_gf256_mul_inv_round_trip(void) {
    aether_gf256_init();
    for (int a = 1; a < 256; a++) {
        uint8_t inv = aether_gf256_inv((uint8_t)a);
        uint8_t one = aether_gf256_mul((uint8_t)a, inv);
        assert(one == 1);
    }
    printf("  PASS test_gf256_mul_inv_round_trip\n");
}

static void test_gf256_mul_commutativity(void) {
    aether_gf256_init();
    for (int a = 0; a < 256; a++) {
        for (int b = 0; b < 256; b++) {
            assert(aether_gf256_mul((uint8_t)a, (uint8_t)b) ==
                   aether_gf256_mul((uint8_t)b, (uint8_t)a));
        }
    }
    printf("  PASS test_gf256_mul_commutativity\n");
}

static void test_gf256_mul_distributivity(void) {
    // a*(b+c) == a*b + a*c
    aether_gf256_init();
    uint8_t a = 0x53, b = 0xCA, c = 0x71;
    uint8_t lhs = aether_gf256_mul(a, aether_gf256_add(b, c));
    uint8_t rhs = aether_gf256_add(aether_gf256_mul(a, b), aether_gf256_mul(a, c));
    assert(lhs == rhs);
    printf("  PASS test_gf256_mul_distributivity\n");
}

// ── RlncEncoder ───────────────────────────────────────────────────────────────

static void test_encoder_systematic_first_k_packets(void) {
    int k = 4, sym = 8;
    uint8_t **src = make_source(k, sym);
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, true);
    assert(enc != NULL);

    uint8_t *coeff = (uint8_t *)malloc((size_t)k);
    uint8_t *data  = (uint8_t *)malloc((size_t)sym);
    for (int i = 0; i < k; i++) {
        aether_rlnc_encoder_next_packet(enc, coeff, data);
        // Coefficient vector must be e_i.
        assert(coeff[i] == 1);
        for (int j = 0; j < k; j++) {
            if (j != i) assert(coeff[j] == 0);
        }
        // Data must equal source[i].
        assert(memcmp(data, src[i], (size_t)sym) == 0);
    }
    free(coeff);
    free(data);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_encoder_systematic_first_k_packets\n");
}

static void test_encoder_repair_packets_not_all_zero(void) {
    int k = 3, sym = 4;
    uint8_t **src = make_source(k, sym);
    // Non-systematic encoder — every packet is a repair packet.
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, false);
    assert(enc != NULL);

    uint8_t *coeff = (uint8_t *)malloc((size_t)k);
    uint8_t *data  = (uint8_t *)malloc((size_t)sym);
    for (int i = 0; i < 20; i++) {
        aether_rlnc_encoder_next_packet(enc, coeff, data);
        bool any_nonzero = false;
        for (int j = 0; j < k; j++) {
            if (coeff[j] != 0) { any_nonzero = true; break; }
        }
        assert(any_nonzero && "repair packet has all-zero coefficient vector");
    }
    free(coeff);
    free(data);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_encoder_repair_packets_not_all_zero\n");
}

// ── RlncDecoder ───────────────────────────────────────────────────────────────

static void test_decoder_round_trip_k4(void) {
    int k = 4, sym = 8;
    uint8_t **src = make_source(k, sym);
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, true);
    aether_rlnc_decoder_t *dec = aether_rlnc_decoder_new(k, sym);
    assert(enc && dec);

    uint8_t *coeff = (uint8_t *)malloc((size_t)k);
    uint8_t *data  = (uint8_t *)malloc((size_t)sym);
    while (!aether_rlnc_decoder_is_complete(dec)) {
        aether_rlnc_encoder_next_packet(enc, coeff, data);
        aether_rlnc_decoder_add_packet(dec, coeff, data);
    }
    free(coeff);
    free(data);

    size_t decoded_len = 0;
    uint8_t *decoded = aether_rlnc_decoder_try_decode(dec, &decoded_len);
    assert(decoded != NULL);
    assert(decoded_len == k * sym);

    for (int i = 0; i < k; i++) {
        assert(memcmp(decoded + i * sym, src[i], (size_t)sym) == 0);
    }
    free(decoded);
    aether_rlnc_decoder_free(dec);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_decoder_round_trip_k4\n");
}

static void test_decoder_exactly_k_systematic_complete(void) {
    int k = 3, sym = 4;
    uint8_t **src = make_source(k, sym);
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, true);
    aether_rlnc_decoder_t *dec = aether_rlnc_decoder_new(k, sym);

    uint8_t *coeff = (uint8_t *)malloc((size_t)k);
    uint8_t *data  = (uint8_t *)malloc((size_t)sym);
    for (int i = 0; i < k; i++) {
        aether_rlnc_encoder_next_packet(enc, coeff, data);
        aether_rlnc_decoder_add_packet(dec, coeff, data);
    }
    free(coeff);
    free(data);

    assert(aether_rlnc_decoder_is_complete(dec));
    assert(aether_rlnc_decoder_rank(dec) == k);

    aether_rlnc_decoder_free(dec);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_decoder_exactly_k_systematic_complete\n");
}

static void test_decoder_linearly_dependent_packet_ignored(void) {
    int k = 2, sym = 4;
    uint8_t **src = make_source(k, sym);
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, true);
    aether_rlnc_decoder_t *dec = aether_rlnc_decoder_new(k, sym);

    uint8_t coeff0[2], data0[4];
    aether_rlnc_encoder_next_packet(enc, coeff0, data0);
    bool added1 = aether_rlnc_decoder_add_packet(dec, coeff0, data0);
    int rank_before = aether_rlnc_decoder_rank(dec);
    bool added2 = aether_rlnc_decoder_add_packet(dec, coeff0, data0); // duplicate
    (void)added1; (void)added2;
    assert(aether_rlnc_decoder_rank(dec) == rank_before &&
           "duplicate packet should not increase rank");

    aether_rlnc_decoder_free(dec);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_decoder_linearly_dependent_packet_ignored\n");
}

static void test_decoder_is_complete_at_rank_k(void) {
    int k = 2, sym = 3;
    uint8_t **src = make_source(k, sym);
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, true);
    aether_rlnc_decoder_t *dec = aether_rlnc_decoder_new(k, sym);

    uint8_t *c = (uint8_t *)malloc((size_t)k);
    uint8_t *d = (uint8_t *)malloc((size_t)sym);

    assert(!aether_rlnc_decoder_is_complete(dec));
    aether_rlnc_encoder_next_packet(enc, c, d); aether_rlnc_decoder_add_packet(dec, c, d);
    assert(!aether_rlnc_decoder_is_complete(dec));
    aether_rlnc_encoder_next_packet(enc, c, d); aether_rlnc_decoder_add_packet(dec, c, d);
    assert(aether_rlnc_decoder_is_complete(dec));

    free(c); free(d);
    aether_rlnc_decoder_free(dec);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_decoder_is_complete_at_rank_k\n");
}

static void test_decoder_repair_only_round_trip(void) {
    int k = 4, sym = 8;
    uint8_t **src = make_source(k, sym);
    // Non-systematic encoder — all repair packets.
    aether_rlnc_encoder_t *enc = aether_rlnc_encoder_new(
        (const uint8_t * const *)src, k, sym, false);
    aether_rlnc_decoder_t *dec = aether_rlnc_decoder_new(k, sym);

    uint8_t *c = (uint8_t *)malloc((size_t)k);
    uint8_t *d = (uint8_t *)malloc((size_t)sym);
    int attempts = 0;
    while (!aether_rlnc_decoder_is_complete(dec)) {
        aether_rlnc_encoder_next_packet(enc, c, d);
        aether_rlnc_decoder_add_packet(dec, c, d);
        assert(++attempts < 200 && "repair-only decoder stalled");
    }
    free(c); free(d);

    size_t decoded_len = 0;
    uint8_t *decoded = aether_rlnc_decoder_try_decode(dec, &decoded_len);
    assert(decoded != NULL);
    assert(decoded_len == k * sym);
    for (int i = 0; i < k; i++) {
        assert(memcmp(decoded + i * sym, src[i], (size_t)sym) == 0);
    }
    free(decoded);
    aether_rlnc_decoder_free(dec);
    aether_rlnc_encoder_free(enc);
    free_source(src, k);
    printf("  PASS test_decoder_repair_only_round_trip\n");
}

// ── RlncCodec (vtable) ────────────────────────────────────────────────────────

static void test_codec_k1_single_symbol(void) {
    aether_rlnc_codec_t *codec = aether_rlnc_codec_new(1);
    assert(codec != NULL);
    aether_fec_codec_t *base = (aether_fec_codec_t *)codec;

    const uint8_t source[] = {0xDE, 0xAD, 0xBE, 0xEF};
    size_t source_len = sizeof(source);
    int target_count = 2;
    size_t encoded_len = 0;

    uint8_t *encoded = base->encode(base, source, source_len, target_count, &encoded_len);
    assert(encoded != NULL);

    // Build packet-pointer and symbol-length arrays.
    size_t pkt_size = encoded_len / (size_t)target_count;
    const uint8_t **pkts    = (const uint8_t **)malloc((size_t)target_count * sizeof(uint8_t *));
    size_t         *lengths = (size_t *)malloc((size_t)target_count * sizeof(size_t));
    for (int i = 0; i < target_count; i++) {
        pkts[i]    = encoded + i * pkt_size;
        lengths[i] = pkt_size;
    }

    size_t decoded_len = 0;
    uint8_t *decoded = base->try_decode(base, pkts, lengths, target_count, 1, &decoded_len);
    assert(decoded != NULL);
    assert(decoded_len >= source_len);
    assert(memcmp(decoded, source, source_len) == 0);

    free(decoded);
    free(lengths);
    free(pkts);
    free(encoded);
    aether_rlnc_codec_free(codec);
    printf("  PASS test_codec_k1_single_symbol\n");
}

static void test_codec_large_payload_round_trip(void) {
    int k = 16;
    aether_rlnc_codec_t *codec = aether_rlnc_codec_new(k);
    assert(codec != NULL);
    aether_fec_codec_t *base = (aether_fec_codec_t *)codec;

    // 1024-byte source.
    size_t source_len = 1024;
    uint8_t *source = (uint8_t *)malloc(source_len);
    for (size_t i = 0; i < source_len; i++) source[i] = (uint8_t)(i & 0xFF);

    int    target_count = 20; // 16 systematic + 4 repair
    size_t encoded_len  = 0;
    uint8_t *encoded = base->encode(base, source, source_len, target_count, &encoded_len);
    assert(encoded != NULL);

    size_t          pkt_size = encoded_len / (size_t)target_count;
    const uint8_t **pkts    = (const uint8_t **)malloc((size_t)target_count * sizeof(uint8_t *));
    size_t         *lengths = (size_t *)malloc((size_t)target_count * sizeof(size_t));
    for (int i = 0; i < target_count; i++) {
        pkts[i]    = encoded + i * pkt_size;
        lengths[i] = pkt_size;
    }

    size_t decoded_len = 0;
    uint8_t *decoded = base->try_decode(base, pkts, lengths, target_count, k, &decoded_len);
    assert(decoded != NULL);
    assert(decoded_len >= source_len);
    assert(memcmp(decoded, source, source_len) == 0);

    free(decoded);
    free(lengths);
    free(pkts);
    free(encoded);
    free(source);
    aether_rlnc_codec_free(codec);
    printf("  PASS test_codec_large_payload_round_trip\n");
}

// ── main ──────────────────────────────────────────────────────────────────────

int main(void) {
    printf("=== RLNC Tests ===\n");

    // GF(256) arithmetic
    test_gf256_add_is_xor();
    test_gf256_mul_by_zero();
    test_gf256_mul_by_one();
    test_gf256_mul_inv_round_trip();
    test_gf256_mul_commutativity();
    test_gf256_mul_distributivity();

    // Encoder
    test_encoder_systematic_first_k_packets();
    test_encoder_repair_packets_not_all_zero();

    // Decoder
    test_decoder_round_trip_k4();
    test_decoder_exactly_k_systematic_complete();
    test_decoder_linearly_dependent_packet_ignored();
    test_decoder_is_complete_at_rank_k();
    test_decoder_repair_only_round_trip();

    // Codec vtable
    test_codec_k1_single_symbol();
    test_codec_large_payload_round_trip();

    printf("=== All RLNC tests passed ===\n");
    return 0;
}

File: python/tests/test_rlnc.py
# SPDX-License-Identifier: MIT
"""Unit tests for the RLNC engine (GF(2⁸), encoder, decoder, codec)."""

from __future__ import annotations

import unittest

from aethernet.transport.rlnc import (
    RlncCodec,
    RlncDecoder,
    RlncEncoder,
    gf256_add,
    gf256_inv,
    gf256_mul,
)


# ── GF(2⁸) arithmetic ─────────────────────────────────────────────────────────

class TestGf256Arithmetic(unittest.TestCase):

    def test_add_is_xor(self):
        self.assertEqual(gf256_add(0xAB, 0xCD), 0xAB ^ 0xCD)
        self.assertEqual(gf256_add(0, 0), 0)

    def test_mul_by_zero(self):
        for v in range(256):
            self.assertEqual(gf256_mul(v, 0), 0)
            self.assertEqual(gf256_mul(0, v), 0)

    def test_mul_by_one(self):
        for v in range(1, 256):
            self.assertEqual(gf256_mul(v, 1), v)
            self.assertEqual(gf256_mul(1, v), v)

    def test_mul_inv_round_trip(self):
        """Mul(a, Inv(a)) == 1 for all non-zero a."""
        for v in range(1, 256):
            self.assertEqual(gf256_mul(v, gf256_inv(v)), 1,
                             f"Mul({v}, Inv({v})) != 1")

    def test_mul_commutativity(self):
        for a in range(1, 32):
            for b in range(1, 32):
                self.assertEqual(gf256_mul(a, b), gf256_mul(b, a))

    def test_mul_distributivity(self):
        """a*(b+c) == a*b + a*c  (spot-check)."""
        a, b, c = 0x53, 0xCA, 0x77
        lhs = gf256_mul(a, gf256_add(b, c))
        rhs = gf256_add(gf256_mul(a, b), gf256_mul(a, c))
        self.assertEqual(lhs, rhs)

    def test_inv_of_one(self):
        self.assertEqual(gf256_inv(1), 1)

    def test_inv_raises_for_zero(self):
        with self.assertRaises(ZeroDivisionError):
            gf256_inv(0)


# ── RlncEncoder ───────────────────────────────────────────────────────────────

class TestRlncEncoder(unittest.TestCase):

    def _make_encoder(self, k: int, sym_size: int = 3, systematic: bool = True):
        source = [bytearray(range(i, i + sym_size)) for i in range(k)]
        return RlncEncoder(source, systematic=systematic), source

    def test_systematic_first_k_packets_identity_coeff(self):
        k = 4
        enc, source = self._make_encoder(k)
        for i in range(k):
            coeff, data = enc.next_packet()
            # Coefficient vector must be e_i.
            self.assertEqual(len(coeff), k)
            for j in range(k):
                self.assertEqual(coeff[j], 1 if j == i else 0,
                                 f"pkt {i}: coeff[{j}] should be {1 if j==i else 0}")
            # Data must equal source symbol.
            self.assertEqual(bytes(data), bytes(source[i]),
                             f"systematic pkt {i} data mismatch")

    def test_repair_packets_not_all_zero(self):
        enc, _ = self._make_encoder(k=3, systematic=False)
        for i in range(20):
            coeff, _ = enc.next_packet()
            self.assertFalse(all(c == 0 for c in coeff),
                             f"repair pkt {i} has all-zero coefficient vector")

    def test_error_on_empty_source(self):
        with self.assertRaises(ValueError):
            RlncEncoder([])


# ── RlncDecoder ───────────────────────────────────────────────────────────────

class TestRlncDecoder(unittest.TestCase):

    def test_starts_at_rank_zero(self):
        dec = RlncDecoder(k=4, symbol_size=8)
        self.assertEqual(dec.rank, 0)
        self.assertFalse(dec.is_complete)

    def test_linearly_dependent_packet_rejected(self):
        dec = RlncDecoder(k=3, symbol_size=4)
        coeff = bytes([1, 0, 0])
        data  = bytes([10, 20, 30, 40])
        self.assertTrue(dec.add_packet(coeff, data))
        self.assertFalse(dec.add_packet(coeff, data),
                         "duplicate packet must not increase rank")
        self.assertEqual(dec.rank, 1)

    def test_complete_after_k_independent_packets(self):
        k, s = 3, 2
        dec = RlncDecoder(k=k, symbol_size=s)
        for i in range(k):
            coeff = bytearray(k)
            coeff[i] = 1
            data = bytes([i + 1, i + 100])
            dec.add_packet(bytes(coeff), data)
        self.assertTrue(dec.is_complete)

    def test_try_decode_returns_none_when_incomplete(self):
        dec = RlncDecoder(k=4, symbol_size=4)
        self.assertIsNone(dec.try_decode())

    def test_try_decode_symbol_ordering(self):
        k, s = 3, 2
        dec = RlncDecoder(k=k, symbol_size=s)
        sources = [(0xAA, 0xBB), (0xCC, 0xDD), (0xEE, 0xFF)]
        for i, (a, b) in enumerate(sources):
            coeff = bytearray(k)
            coeff[i] = 1
            dec.add_packet(bytes(coeff), bytes([a, b]))
        result = dec.try_decode()
        self.assertIsNotNone(result)
        for i, (a, b) in enumerate(sources):
            self.assertEqual(result[i * s],     a, f"symbol[{i}][0]")
            self.assertEqual(result[i * s + 1], b, f"symbol[{i}][1]")


# ── RlncCodec: round-trips ────────────────────────────────────────────────────

class TestRlncCodecRoundTrip(unittest.TestCase):

    def _decode_split(self, encoded: bytes, k: int, count: int):
        pkt_size = len(encoded) // count
        return [encoded[i * pkt_size:(i + 1) * pkt_size] for i in range(count)]

    def test_k4_systematic_decode(self):
        source = b"aether-rlnc-python-test-k4"
        codec  = RlncCodec(generation_size=4)
        encoded = codec.encode(source, 4)
        pkts = self._decode_split(encoded, 4, 4)
        decoded = codec.try_decode(pkts, 4)
        self.assertIsNotNone(decoded)
        self.assertEqual(decoded[:len(source)], source)

    def test_k4_with_repair_overhead(self):
        source = b"round-trip with extra repair packets"
        codec  = RlncCodec(generation_size=4)
        encoded = codec.encode(source, 6)
        pkts = self._decode_split(encoded, 4, 6)[:4]  # just systematic
        decoded = codec.try_decode(pkts, 4)
        self.assertIsNotNone(decoded)
        self.assertEqual(decoded[:len(source)], source)

    def test_repair_only_decode(self):
        source = b"repair-only round-trip in Python"
        codec  = RlncCodec(generation_size=4)
        encoded = codec.encode(source, 8)
        pkts = self._decode_split(encoded, 4, 8)[4:]  # skip systematic
        decoded = codec.try_decode(pkts, 4)
        self.assertIsNotNone(decoded, "repair-only decode failed")
        self.assertEqual(decoded[:len(source)], source)

    def test_k1_single_symbol(self):
        source = b"z"
        codec  = RlncCodec(generation_size=1)
        encoded = codec.encode(source, 2)
        pkts = self._decode_split(encoded, 1, 2)[:1]
        decoded = codec.try_decode(pkts, 1)
        self.assertIsNotNone(decoded)
        self.assertEqual(decoded[0], ord("z"))

    def test_k16_large_payload(self):
        source = bytes(range(256)) * 4  # 1024 bytes
        codec  = RlncCodec(generation_size=16)
        encoded = codec.encode(source, 20)
        pkts = self._decode_split(encoded, 16, 20)
        decoded = codec.try_decode(pkts, 16)
        self.assertIsNotNone(decoded)
        self.assertEqual(decoded[:len(source)], source)

    def test_empty_received_returns_none(self):
        codec = RlncCodec(generation_size=4)
        self.assertIsNone(codec.try_decode([], 4))


# ── Codec metadata ─────────────────────────────────────────────────────────────

class TestRlncCodecMetadata(unittest.TestCase):

    def test_codec_name(self):
        self.assertEqual(RlncCodec().codec_name, "RLNC-GF256")

    def test_overhead_fraction(self):
        self.assertAlmostEqual(RlncCodec().overhead_fraction, 0.05)

    def test_rejects_generation_size_zero(self):
        with self.assertRaises(ValueError):
            RlncCodec(generation_size=0)

    def test_rejects_generation_size_256(self):
        with self.assertRaises(ValueError):
            RlncCodec(generation_size=256)


if __name__ == "__main__":
    unittest.main()
